src/initiallystoppedprocess.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+    SPDX-FileCopyrightText: Zeno Endemann <zeno.endemann@kdab.com>
+    SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#include "initiallystoppedprocess.h"
+
+#include <QDebug>
+#include <QFile>
+#include <QLoggingCategory>
+#include <QSocketNotifier>
+#include <QStandardPaths>
+
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <vector>
+
+namespace {
+Q_LOGGING_CATEGORY(initiallystoppedprocess, "hotspot.initiallystoppedprocess")
+
+void sendSignal(pid_t pid, int signal)
+{
+    if (!kill(pid, signal)) {
+        qCCritical(initiallystoppedprocess, "Failed to send %u to %u", signal, pid);
+    }
+}
+}
+
+InitiallyStoppedProcess::~InitiallyStoppedProcess()
+{
+    kill();
+}
+
+bool InitiallyStoppedProcess::createProcessAndStop(const QString& exePath, const QStringList& exeOptions,
+                                                   const QString& workingDirectory)
+{
+    kill();
+
+    // convert arguments and working dir into what the C API needs
+
+    std::vector<QByteArray> args;
+    args.reserve(exeOptions.size() + 1);
+    args.emplace_back(exePath.toLocal8Bit());
+    for (const auto& opt : exeOptions)
+        args.emplace_back(opt.toLocal8Bit());
+    const auto wd = workingDirectory.toLocal8Bit();
+
+    std::vector<char*> argsArray(args.size() + 1);
+    for (size_t i = 0; i < args.size(); ++i)
+        argsArray[i] = args[i].data();
+    argsArray.back() = nullptr;
+
+    // fork
+    m_pid = fork();
```

Comment:
@milianw 
> forking means we COW the whole address space which can be really slow when that is large. that would e.g. be the case when you start hotspot, analyze some large file and then rerun the record from there.
>
> is there a way this could use QProcess such that we can leverage vfork/posix_spawn when available? or is > there maybe something like that in boost which we could use to get this behavior in a more efficient manner?
> 
> see also: https://codereview.qt-project.org/c/qt/qtbase/+/417829

@zeno-endemann-kdab 
> sure this could be improved. Don't know what's available in boost for that. vfork seems a bit dangerous to me, for posix_spawn I suppose one would need to move this initially-stopped-exec functionality in a separate helper binary or something.

@lievenhey 
> we can only use vfork if we create a new binary, since vfork only allows for __exit() and exec* calls. I think clone with CLONE_WM | CLONE_VFORK might be the easiest way to solve this.

@zeno-endemann-kdab 
> From my understanding, VFORK suspends the calling process until exec is called, which we do not want here at all. I don't think VFORK is usable for this. posix_spawn and a helper process should work though.

Reply:
I think vfork is fine, why shouldn't we be halting the process until exec is called? that should be done quickly after all.

Reply:
Maybe I misunderstand something, but we fork here to immediately stop the child before exec-ing so that the parent process has time to attach perf. Making the parent wait for the child to exec would obviously deadlock then.

Reply:
from the docs:
> vfork() differs from [fork(2)](https://man7.org/linux/man-pages/man2/fork.2.html) in that the calling thread is
       suspended until the child terminates (either normally, by calling
       [_exit(2)](https://man7.org/linux/man-pages/man2/_exit.2.html), or abnormally, after delivery of a fatal signal), or it
       makes a call to [execve(2)](https://man7.org/linux/man-pages/man2/execve.2.html).

`SIGSTOP` is not a fatal signal

Reply:
One can of course create a helper binary ("exec_stopped" or something) that contains the sigstop-then-exec logic, and that one could be vfork'ed safely, if that is what you meant.

Reply:
indeed, my bad. anyhow, let's ignore this for now and I'll try to look into alternatives to `fork` in the future